### ISSUE TITLE
Persistence of semagrams in Jupyter notebooks

### ISSUE THREAD

**Issue by epatters** [AlgebraicJulia/Semagrams.jl#9]
Suppose you have created a semagram in a Jupyter notebook, saved it, closed the notebook, and shut down the kernel. When you reopen the notebook and rerun the cells, the semagram is lost.

It seems very important to be able to persist semagrams across notebook runs. I don't know exactly how this should work, but I assume it could be done by storing a JSON blob inside the notebook.

**Comment by olynch** [comment]
Yes, this is something I've been thinking about. I think there are two problems here.

1. The communication between Semagrams and Julia should be bidirectional, which means that Julia is also going to have to be able to think about Semagram layout.
2. We need a way of, once the data (including layout) has gotten back to Julia, to store it across sessions.

The first problem I understand well, but the second not so much. After all, the values of variables are not stored across notebook runs.

My inclination here is to make saving/loading the semagram *explicit*. That is, we save the semagram to a separate JSON file. This would allow for the semagram to be used in scripts, as well as edited in future Jupyter sessions. What are your thoughts on this?

**Comment by epatters** [comment]
Regarding (2), I think that instead of having Semagrams send the data to Julia and then persist it from within Julia, the Semagrams widget should store the data inside the Jupyter notebook since it, not the Julia kernel, has direct access.

Looking into a bit, ipywidgets have a mechanism to [save their state](https://ipywidgets.readthedocs.io/en/latest/embedding.html#save-notebook-widget-state) as metadata inside the notebook. From a quick glance at the [source code](https://github.com/jupyter-widgets/ipywidgets/blob/f839d51a77c905e1462c2e9a7d02d63586341a67/jupyterlab_widgets/src/manager.ts#L526), it just stores an arbitrary JSON blob in the notebook. Presumably this would be done with cell metadata rather than the notebook metadata. See the [schema](https://nbformat.readthedocs.io/en/latest/format_description.html#metadata) for metadata in Jupyter notebooks.

What do you think?

**Comment by olynch** [comment]
So, this is an interesting idea, and I will look into it. I think this will ultimately depend on whether persistence is handled in WebIO; if there's a persistence story in this way for WebIO, then we can take advantage of it, but otherwise we're going to have to work at a far lower level than we are currently working at in order to get this to work, and it would potentially take a PR to WebIO.

**Comment by olynch** [comment]
It seems to me that for ipywidgets, the primary purpose of embedding widgets within the notebook is to be able to export the notebook. I don't think that the connection between an ipywidget and the python kernel can be saved between restarts; the widget would stay in place until you reran the line of code that produced it, at which point it would be refreshed to its original state.

I think that saving semagrams data to an external file would be preferable because the external file could then be used in a non-Jupyter notebook context, such as the Mux backend for Semagrams, or in an actual library. If the semagrams are tied to the jupyter notebook, then if you want to use them in anything else you would have to manually copy them, which defeats the point.

**Comment by olynch** [comment]
With semagrams next, the thing that I think would make this issue "solved" is a coherent serialization story for semagrams, which shouldn't be too hard.
